Add PQC blog post. (#49)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2025-12-12-pqc-on-cheriot.markdown

These algorithms are referred to as post-quantum cryptography (PQC).
Since NIST standardised a few such algorithms, there's been a lot more interest in seeing them in production, so I spent some time getting the implementations from the Linux Foundation's PQ Code Package to run on CHERIoT.
A lot of companies are building hardware to accelerate these operations, so it seemed useful to have a performance baseline on the CHERIoT Ibex, as well as something that can be used in future CHERIoT-based products.

What are ML-KEM and ML-DSA for?
-------------------------------

I am not a mathematician and so I'm not going to try to explain how these algorithms work, but I am going to explain what they're *for*.

Module-Lattice-Based Key-Encapsulation Mechanism (ML-KEM) is, as the name suggests, an algorithm for key encapsulation.
One side holds a public key and uses it (plus some entropy source) to generate a secret in both plain and encapsulated forms.
The encapsulated secret can be sent to a remote party who holds the corresponding private key.

> The receiver can then recover unencrypted version of the secret (and detect tampering).

Reviewer: ```suggestion
The receiver can then recover the unencrypted version of the secret (and detect tampering).
```

Now, both parties have the same secret and can use it with some key-derivation function to produce something like an AES key for future communication.

Note that this is somewhat more restrictive than traditional key-exchange protocols.

> You don't get to exchange an arbitrary value, the generation step is part of encapsulation.

Reviewer: ```suggestion
You don't get to exchange an arbitrary value: the generation step is part of encapsulation.
```

This also means that it's a fixed size, defined by the algorithm, which is why you typically feed it into a key-derivation function rather than using it directly.

Module-Lattice Digital Signature Algorithm (ML-DSA) has a similarly informative name.
It is intended for providing and validating digital signatures.
It takes a private key, an arbitrary-sized document and context, and produces a signature.
A holder of the associated public key can then validate that the document matches the version signed with the private key and context.

These are both quite low-level building blocks for higher-level protocols.
For example, TLS can use ML-KEM for key exchange and ML-DSA for certificate validation, but also incorporates traditional algorithms in case the PQC algorithms have unexpected weaknesses against classical computers.

Initial porting
---------------

As is usually the case for CHERIoT, porting the C implementations of ML-KEM and ML-DSA required no code changes.
I worked with upstream to slightly simplify the platform-integration layer, so we just provide a single header describing the port.

> For example, the [port heaer for ML-DSA](https://github.com/CHERIoT-Platform/cheriot-pqc/blob/main/include/mldsa_native_config.h) configures the build to produce ML-DSA44 support, defines a custom function for zeroing memory and getting entropy, and adds the `__cheriot_libcall` attribute to the all exported APIs (so we can build them as shared libraries, rather than embedded in a single compartment).

Reviewer: ```suggestion
For example, the [port here for ML-DSA](https://github.com/CHERIoT-Platform/cheriot-pqc/blob/main/include/mldsa_native_config.h) configures the build to produce ML-DSA44 support, defines a custom function for zeroing memory and getting entropy, and adds the `__cheriot_libcall` attribute to the all exported APIs (so we can build them as shared libraries, rather than embedded in a single compartment).
```

The [file for ML-KEM](https://github.com/CHERIoT-Platform/cheriot-pqc/blob/main/include/mlkem_native_config.h) is almost identical.

With these defined, it is possible to build both libraries as CHERIoT shared libraries.
This motivated a bit of cleanup.

> We have a device interface for entropy sources, but it wasn't implemented on the Sail model (which doesn't have an entropy source).

Reviewer: ```suggestion
We have a device interface for entropy sources, but it wasn't implemented for the Sail model (which doesn't have an entropy source).
```

It has a way of exposing the fact that entropy is insecure, so that wasn't a problem, it just needed doing, so I refactored all of the insecure entropy-source drivers to use a common base.
Most encryption algorithms want an API that fills a buffer with entropy.
It's nice if these don't all need to touch the driver directly, so I created a compartment that provides this API and exposes it.

> Now, both libraries are simply consumers of this API.

Reviewer: ```suggestion
Now, both crypto libraries are simply consumers of this API.
```

This also makes it easier to add stateful whitening for entropy drivers for hardware entropy sources that don't do the whitening in hardware.

Most CHERIoT stacks are on the order of 1-2 KiBs.
The PQC algorithms use much more space.
More, in fact, than we permitted.

The previous limitation was based on the precision of bounds rounding.
A CHERI capability compresses the bounds representation by taking advantage of the fact that, for a pointer to an allocation, there is a lot of redundancy between the address of the pointer, the address of the end of the allocation (the top), and the address of the start of the allocation (the base).
The distance from the address to base and top are stored as floating-point values with a shared exponent.
In practical terms, this means that the larger an allocation is, the more strongly aligned its start and end addresses must be.

> The same restrictions apply for any capability that grants access to less than an entire object.

Reviewer: Perhaps make this a footnote?

When you call a function in another compartment, the switcher will truncate the stack capability so that the callee sees only the bit of the stack that you weren't using.
The top and base of the stack must be 16-byte aligned (as an ABI requirement), but a very large stack may have hardware requirements for greater alignment and so may require a gap between the bottom of the caller's stack and the top of the callee's.

Fortunately, we'd added an instruction precisely for this kind of use case: `CSetBoundsRoundDown`.
This takes a capability and a length and truncates it to *at most* that length.
It was a fairly small tweak to the switcher to make it do this, and a much larger amount of time with SMT solvers to convince ourselves that this was a safe thing to do.

This also showed up a bug in our linker's handling of the `CAPALIGN` directive, which rounds a section's base and size up to the required alignment to be representable.
This was not working for sections that followed an explicit alignment directive.
Our stacks must be both at least 16-byte aligned *and* representable as capabilities.
This is now fixed.

> So now we support stacks up to almost 64 KiB.

Reviewer: Might be worth mentioning the reason for the new limit?

> We could easily increase this limit but 64 KiB ought to be enough for anyone.

Reviewer: I know this is a bit flippant, but I think you could do with an argument for why this is the case e.g. can usually use dynamic memory or globals as an alternative.

The output on the CHERIoT Ibex verilator simulation is:

```
PQC benchmark: Starting: stack used: 224 bytes, cycles elapsed: 41
PQC benchmark: Generated ML-KEM key pair: stack used: 14304 bytes, cycles elapsed: 5143987
PQC benchmark: Encrypted secret pair with ML-KEM: stack used: 17440 bytes, cycles elapsed: 1773235
PQC benchmark: Decrypted secret pair with ML-KEM: stack used: 18464 bytes, cycles elapsed: 2176226
PQC benchmark: Compared results successfully for ML-KEM: stack used: 224 bytes, cycles elapsed: 414
PQC benchmark: Generated ML-DSA key pair: stack used: 46912 bytes, cycles elapsed: 3622132
PQC benchmark: Signed message with ML-DSA: stack used: 60544 bytes, cycles elapsed: 5391177
PQC benchmark: Verified message signature with ML-DSA: stack used: 44672 bytes, cycles elapsed: 3674071
PQC benchmark: Correctly failed to verify message signature with ML-DSA after tampering: stack used: 44672 bytes, cycles elapsed: 3673706
```

The ML-KEM encrypt (generate shared secret and encrypted version) and decrypt (recover shared secret from encrypted version) each use around 18 KiB of stack and run in around two million cycles.

> The CHERIoT Ibex should scale up to 200-300 MHz (though may be clocked lower for power reasons in some deployments), but even at 100 MHz that's 50 encryption or decryption operations.

Reviewer: ```suggestion
CHERIoT Ibex should scale up to 200-300 MHz (though may be clocked lower for power reasons in some deployments), but even at 100 MHz that's 50 encryption or decryption operations per second.
```

Remember that this is an operation that typically happens when you establish a connection, then you use a stream cypher such as AES with the exchanged key.

The ML-DSA operations are slower and use a *lot* more stack space (almost 60 KiB for signing!).

> But, even there, the performance is reasonable, under 4 M cycles.

Reviewer: ```suggestion
But, even there, the performance is reasonable: under 4M cycles.
```

Author: This one annoys me, but the space is required from the SI prefix typography rules.

Glitch injection takes the chip out of its specified power or frequency (or thermal) envelope and attempts to introduce bit flips, which can corrupt state in such a way that tamper with signing or leak a key.
These are also effectively impossible to mitigate in software because the software that's attempting the mitigation is vulnerable to the same glitches.
There are some compiler techniques that can make these harder, but they come with a high performance cost.

If power analysis and glitch injection are part of your threat model, the software implementations are not sufficient.
In this case you may also need to worry about someone removing the top of the chip and using a scanning-tunnelling electron microscope to read bits from non-volatile memory.
This used to require tens of thousands of dollars but is now much cheaper.
Devices that need to worry about this often have tiny explosive charges in the package to destroy the chip in cases of tampering.
If that's your threat model, hardware PQC implementations may not be sufficient, at least alone.

> But if you care about attackers on the network being unable to compromise the security of the class of devices, even if they have a magical and imaginary quantum computer, then these should be sufficient.

Reviewer: ```suggestion
But if you care about attackers on the network being unable to compromise the security of the class of devices, even if they have a magical and imaginary quantum computer, then these software implementations should be sufficient.
```
